### ISSUE TITLE
How to tag restaurants that accept Lightning via Qerko in Czechia?

### ISSUE THREAD

**Issue by anton-isidore** [teambtcmap/btcmap-data#4283]
In Czechia there is a service called Querko (https://www.qerko.com) that allows you to checkout (pay) what you have at the table via app without the need to interact with waiters.

They now accept payments via Lightning.

Article about it in both Czech and English:  https://qerko.freshdesk.com/support/solutions/articles/80001077943-how-to-pay-with-bitcoin-jak-zaplatit-bitcoinem

Here is a technical info how they do it (in czech, but Google Translate works):  https://gist.github.com/pepakriz/f7b5f1e8f75b176a1f117569fed462da

They are still kinda in beta, but this brings a discussion about how tag restaurants that have Querko.

- Restaurant probably don't even know that you can pay with Bitcoin. Staff people will be definitely oblivious to this fact.
- Restaurants get fiat, Querko uses payment gateway to bridge it.
- **Customers need to use Querko app (download it or use instant app) to get the LN invoice.** This is the main topic to discuss how to cover this situation.

**Comment by dadofsambonzuki** [comment]
Excellent!

So, as I see it - the Querko app is and ordering *and* point-of-sale app. I think the fact that it is essentially a self check-out app that generates a native LN invoice justifies it for inclusion.

There is clearly a need to differentiate with additional tags though as a companion app is needed. I think the `payment` tag is useful here, more info here: https://wiki.openstreetmap.org/wiki/Key:payment.

Something like `payment:qerko=yes` could be used and BTC Map apps could use this to show additional information to users.

Are we able to get a list of Querko merchants?

Thoughts @bubelov @stackingsaunter @secondl1ght ?

**Comment by pepakriz** [comment]
> Are we able to get a list of Querko merchants?

No problem, I can provide the current list and push all changes immediately on each change.

> Something like payment:qerko=yes could be used and BTC Map apps could use this to show additional information to users.

Qerko in the restaurant doesn't necessarily mean that it has BTC ln payment enabled. For now, in Czechia it's true, but we can disable it individually (e.g. based on the restaurant owner's wish ). So maybe it should need nested tags like `payment:qerko:lightning=yes`.

**Comment by dadofsambonzuki** [comment]
We have separate lightning and btc tags.

https://wiki.openstreetmap.org/wiki/Bitcoin

We could have:

```
payment:qerko=yes
currency:XBT=no
```

You could also use the `currency` and `payment` tag for any other options you have.

**Comment by dadofsambonzuki** [comment]
What markets do you serve outside of CZ?

**Comment by secondl1ght** [comment]
This seems like it shouldn't be included, I think it falls under the same category as places that only accept KYC bitcoin. If you have to install a specific app in order to make the bitcoin payment then it's not really interoperable. We don't want to start handling all of these walled garden solutions which is a lot of maintenance and then only leads to user confusion about why a customer needs to install a proprietary app to pay with bitcoin.

**Comment by bubelov** [comment]
Is Querko app open source? Is it available outside of Play Store, where it can be easily tracked and geo-blocked? Does it require registration before being able to generate native LN invoices? If it requires registration, which fields need to be filled?

There are tons of red flags, I don't think those places are eligible for inclusion, but it's hard to say for sure without more details on how it actually works.

**Comment by anton-isidore** [comment]
When I was at Prague on BtcPrague I tried to pay with app.

App is not opensource and requires you to log-in. However, the email given in registration process is NOT checked.

I also heard that you can pay just by scanning the QRcode and it will go through Instant App to do payment, but I havent had chance to try it on site.

Can anyone living at Prague check it?

**Comment by pepakriz** [comment]
First of all, let me recapitulate these facts about Qerko:

- Qerko mainly operates in Czechia, but currently, the company is trying to expand to different countries (Austria, Germany, Slovakia, Hungary)
- Majority of payments are processed in restaurants, but Qerko cooperates with TV businesses too and they are not opposed to any cooperation in other sectors
- BTC payments in Qerko are in the experimental phase and they are available just in CZ and for gastro business for now. The main obstacle is legislation (Qerko has a license from Czech National Bank. In other countries, it uses a little different money flow to satisfy bureaucracy.).
- Qerko mobile apps are closed source, but maybe I can get an official statement from the company about the possibility of using their backend API for alternative/unofficial clients. Personally, I would love it if there is an open-source web alternative.
- Registration is here because of card tokenization, loyalty program, etc. You can provide more details like phone number, name, but these additional data are not required. They just help to minimize 3D secure challenges. Anyway in Qerko we want to open the topic of "anonym users" again. BTC payments don't require user details (when the total price is under the law limit). But still, some kind of settings should be connected with the customer to be able to work with BTC. For example, we need to know LNURL of the user for potential refunds. So in case of anonymous payment, we still need to know where to send the money back when POS rejects the payment.

So I completely understand the concerns mentioned above. Anyway, I'm able to provide all fresh data for tags if the community is interested.

**Comment by hynek-jina** [comment]
How can we move this forward?

**Comment by dadofsambonzuki** [comment]
This is clearly a Lightning native solution enabled by bitcoiners working within Qerko. It's not perfect, but it is a step in the right direction and I think we should do everything we can to amplify this initiative. It's similar to the Pick n Pay solution in South Africa and we added that.

I see it as a self-checkout option that gives you a BOLT11 invoice. KYC is optional should you want additional features.

There are 800 Qerko businesses in Czech that accept bitcoin that nobody knows about. I think we should be shouting loudly about this and places these locations on the map.

My view is we should do this properly with tags in OSM. This makes the data more useful to others and enables us to handle that data more effectively within BTC Map.

**Comment by stackingsaunter** [comment]
Yeah, it would be ashame not to show them at all. We need people to encourage bitcoin usage, even if soultion is not perfect. But If people will not use it it will disappear instead of improving. If I was a bitcoiner living in Czech I'd love to have them on a map

**Comment by anton-isidore** [comment]
Hell yeah. So what must be done in order to get them. 

What is the correct tag? What can I do to add them.

Also @pepakriz is there any endpoint to get list of restaurants we can utilize? I can imagine we can run some synchronizer from time to time.

**Comment by hynek-jina** [comment]
I assume that tag is settled to `payment:qerko=yes`

What about link to some explanation how Querko actually work? Should it be inserted to every Qerko location on OSM or rather generally somewhere on BTC Map level?

**Comment by dadofsambonzuki** [comment]
> I assume that tag is settled to `payment:qerko=yes`

Yes, I think this is the best way to enable max flexibility when dealing with these locations. We may want to quickly remove them in the future if T&Cs change for example.

We can flag these locations as 'companion app required' and perhaps show them slightly differently on the map. We can link to a page on our wiki with more detailed information.

**Comment by pepakriz** [comment]
In my opinion `payment:qerko=yes` is not enough, because there are restaurants that have native BTC payments together with Qerko. So if btcmap wants to show that BTC ln payment is possible just over Qerko, it's not true.

**Comment by hynek-jina** [comment]
@pepakriz You can use more tags in that case. For example:
```
currency:XBT=yes
payment:onchain=no
payment:qerko=yes
payment:lightning=yes
payment:lightning_contactless=no
```

**Comment by pepakriz** [comment]
@hynek-jina aha, I understand it now. `payment:qerko=yes` without `payment:lightning=yes` tag still means that the restaurant accepts BTC LN but just with Qerko, right? Maybe I'd suggest to use `payment:qerko:lightning=yes` instead to make it more clear and ready for possible future enhancements.

**Comment by hynek-jina** [comment]
@pepakriz I don't understand the benefit of `payment:qerko:lightning` compared to `payment:qerko`

**Comment by pepakriz** [comment]
> I don't understand the benefit of payment:qerko:lightning compared to payment:qerko

For example, I can imagine in the future some Qerko restaurants will start to accept BTC over Qerko without currency conversion. So another tag should be introduced. The same for `lightning_contactless` over Qerko. I cannot predict concrete additional tags, but It seems right to me to introduce `payment:qerko` tag as a namespace and add concrete tags under this namespace.

**Comment by dadofsambonzuki** [comment]
I think ```payment:qerko:lightning``` is better, because it is more specific.

Other users of OSM may just pick-up the XBT and Lightning tags without knowledge of the Qerko dependency and come to the wrong conclusion.

In order to make [automated edits](https://wiki.openstreetmap.org/wiki/Automated_edits) to OSM, you will need to consult with the Czech OSM mods and discuss your change else your changeset will likely be reverted.

I think we should also open a discussion on the [OSM Wiki Payment Page](https://wiki.openstreetmap.org/wiki/Talk:Key:payment:*) to introduce the Qerko tag before editing the main page.

**Comment by bubelov** [comment]
Automated edits are extremely toxic, it's never a good idea to do it without consulting with the local OSM community and getting an explicit approval.

We should also make sure that:
- The data source is 100% up do date, trusted and reliable
- The import script shouldn't create duplicates, there needs to be a way to check if the place is already on OSM, and it's not as easy as it might sound

> I assume that tag is settled to payment:qerko=yes

Sounds good, we certainly need a way to differentiate native LN and companion app-based payment flows. Perhaps we can add a common tag later, something like `payment:lightning:direct` = `no` && `payment:lightning:provider` == `https://...`

**Comment by bubelov** [comment]
BTC Map itself only shows places which have one of the following tags:

- `currency:XBT` = `yes`
- `payment:bitcoin` = `yes` (there are still ~3,000 places with this legacy tag but it's being phased out)

So, if you want those places to appear on BTC Map, they should also have `currency:XBT` tag. I think it's widely agreed that this tag alone just means that you can pay with bitcoins somehow, and we need to look at the additional tags in order to figure out the details

**Comment by hynek-jina** [comment]
How can we contact Czech OSM mods?

**Comment by bubelov** [comment]
> How can we contact Czech OSM mods?

OSM usually shows the main local comms channels after every edit

https://wiki.openstreetmap.org/wiki/Automated_edits

> The creation and usage of such tools should conform to the [Automated Edits code of conduct](https://wiki.openstreetmap.org/wiki/Automated_Edits_code_of_conduct) and details of edits should be documented on a wiki page in [Category:Automated edits log](https://wiki.openstreetmap.org/wiki/Category:Automated_edits_log). Undiscussed mechanical edits may be reverted without any discussion

OSM is very tough on any automation, I would also consider adding those locations manually, since it would improve the quality of the data. OSM elements aren't simple lat/lon pairs, many stores have a complex geometry and the best way to map places often depends on the context (place type, surroundings, etc)

In my view, the best way to handle that is to have a few dedicated mappers who are willing to pick up OSM mapping skills. Adding places gradually enables early feedback from more experienced local mappers, so it greatly minimizes our chances to break some local conventions and piss off the mods

**Comment by hynek-jina** [comment]
Got it , but it is circa 1000 locations 🤯

**Comment by bubelov** [comment]
How much data about those places can we source? I don't think that the payment company keeps precise geometry of their clients, we would be lucky to have a lat/lon pair, and it's just not enough to automate the import process anyway

We'll also need:
- Working hours
- Contact details (`contact:*` namespace, phone, email, website, socials, etc)
- Amenity (what kind of place is it, and which icon should we display on the pin)

Do we even have this data? If not, we'll probably need to manually fetch it from a bunch of public sources. Please note that Google Maps is an illegal source and we can have a lot of issues if someone detects an import from Google Maps. It's kind of a grey area since it's not easy to prove, but I'd avoid advising people to go there to fetch the missing fields

Maybe we can try focusing on quality instead of quantity and start with the places which are really good and which should be advertised to locals and bitcoiner tourists alike. It's really hard to figure out the best solution without looking into the data that company holds

**Comment by stackingsaunter** [comment]
FYI: I used **payment:querko:lightning** tag here https://www.openstreetmap.org/node/4238247862

**Comment by hynek-jina** [comment]
Thanks @stackingsaunter 
But I think that tag `payment:lightning=no` should be wit `no` (not `yes`, because the restaurant is not able to accept LN payment without Qerko)

**Comment by bubelov** [comment]
Yea, I guess `payment:lightning` can be omitted entirely, since all the `payment:*` tags are optional, the only tag which qualifies OSM elements for inclusion is `currency:XBT`. The web app has some widgets to give user a hint on which Bitcoin payment types are supported, but our mobile apps don't even display that.

Also, `payment:lightning` is a bit controversial, since those payments must go trough the different payment rails (Qerko app). I agree that those places are useful for our users, but I'm not sure we can call them lightning-enabled. Maybe it would be cleaner to only use `payment:lightning` for places which a direct LN wallet support, and use a separate payment tag for each particular "indirect" bitcoin payments provider (which might even use something else instead of LN, our only goal is to make sure the users aren't misinformed).

Another issue we need to address is compliance with OSM verifiability requirements:

https://wiki.openstreetmap.org/wiki/Verifiability

In short, all the tags should be easily verifiable, which means there should be a Bitcoin sign or sticker, or at least an employee who's aware that the shop really accepts bitcoins. Otherwise, those tags will be stripped by other OSM users eventually.

**Comment by stackingsaunter** [comment]
Corrected accordingly

**Comment by anton-isidore** [comment]
I try to push a mit this forward. 

@pepakriz are you please able to provide the data you have mentioned?  With data on hand I believe we can answer many questions about possibility to import data automatically.

**Comment by dadofsambonzuki** [comment]
Galoy worked on a tool that checked for existing locations in OSM before
making a changeset. Let me dig out the link.

Feels like we need to develop something like this more generally if we're
going to onboard millions of merchants and sometimes large groups via PoS
adoption.

**Comment by pepakriz** [comment]
@anton-isidore the data are here: https://gist.githubusercontent.com/pepakriz/571d6a884c76900c5cff28dc4bbf6903/raw/b507c1321417af522f65d2e92ae84fe3eec1f91b/qerko-btcmap.json

**Comment by dadofsambonzuki** [comment]
Cool. Is it possible to get the addition data Igor mentioned:

   - Working hours
   - Contact details (contact:* namespace, phone, email, website, socials,
   etc)
   - Amenity (what kind of place is it, and which icon should we display on
   the pin)

**Comment by bubelov** [comment]
The root array is called "restaurants", so I guess the amenity is known

**Comment by bubelov** [comment]
I checked a few random entries and GPS data seems accurate enough. So, we only need at least a single contact method per entry in order to proceed with the manual import.

Adding emails would be super helpful since it allows us to automate check date bumps. We can send automated emails to merchants which weren't checked for a few months and bump their check date tags or strip all the bitcoin tags, depending on their responses. Import is just the first step, we also need to think how to maintain this dataset.

An automated import still needs to figure out how to avoid creating duplicates. We can probably query all the other elements in 100m radius or something, and see if some names are matching (or almost matching). I have no idea if it's possible and if it's allowed though.

**Comment by dadofsambonzuki** [comment]
"We can probably query all the other elements in 100m radius or something,
and see if some names are matching (or almost matching). I have no idea if
it's possible and if it's allowed though."

This is what the galoy code did.

**Comment by anton-isidore** [comment]
It's 837 entiries, it it maybe doable manually for now. 

Later it would be great if Querko provides API so data can be updated.

Also business idea: it may be even interesting for Querko to do it (maintain data sync) to get more revenew as Bitcoiners will definitely prefer the Querko restaurants before others.

**Comment by bubelov** [comment]
Galoy import was a disaster which almost got us de-platformed and alienated a lot of OSM mods, so apparently that script didn't work as expected =)

**Comment by dadofsambonzuki** [comment]
It was a disaster. The script wasn't the problem, it was the poor quality
data inserted without consultation that was the issue.

**Comment by dadofsambonzuki** [comment]
Here is the [Galoy importer code](https://github.com/vindard/btcmap-galoy-importer).

Probably worth a look to see if it can be of use or can be improved.

**Comment by hynek-jina** [comment]
We have discussed the strategy with @pepakriz yesterday and the plan is following.

1. Add businesses manually first.
2. Store the OSM id to be able to edit them automatically. 
_We will do both steps on few examples first and then organise some hackathon to beat it._
4. Generate list of new businesses on monthly basis and keep adding them manually until we find better solution.

**Comment by bubelov** [comment]
Might be relevant: 

https://github.com/teambtcmap/btcmap-data/issues/4548

We should probably display some kind of a warning when showing places with indirect Bitcoin acceptance

**Comment by hynek-jina** [comment]
What about this kind of disclaimer?
https://btcmap.org/merchant/node:11035285999

**Comment by dadofsambonzuki** [comment]
I think it's best for us to generate the disclaimer based on the
payment:qerko:lightning tag.

Maybe we have an additional icon for those locations that require a
companion app or a slightly different Lightning icon.

**Comment by hynek-jina** [comment]
Just as an activity update.
We are pairing the Qerko businesses with OSM node IDs. There are volunteers helping with that and we will be able to update the businesses automatically afterwards. 

Here are the tagging instructions (in czech):
https://alert-cover-568.notion.site/Tagov-n-Qerko-podnik-v-OSM-be79887c5833484bbf886573b10b384e

People can participate or send donations to other volunteers. Volunteers has covered already over 12% (109 businesses) during today.

**Comment by bubelov** [comment]
Yea, repurposing the `description` tag can cause conflicts, it's better to add a specific or a generic warning, based on the additional tags. The generic one might look like:

`payment:lightning:requires_companion_app` = `yes` or `payment:lightning:direct` = `false`

`payment:qerko:lightning` is also fine, but it won't scale if we're going to have many different data sources requiring different companion apps. Keeping a whitelist of payment providers with their own specific quirks is not the cleanest solution.

We already have an African payments provider which also requires a companion app, and we aren't showing any warnings for those places, so it would be nice to cover all the current and future use cases with a single set of generic tags.

**Comment by dadofsambonzuki** [comment]
I'm not a fan of `payment:lightning:requires_companion_app = yes` or `payment:lightning:direct = false` as they don't give me enough information as a customer.

I think being specific is best (i.e. `payment:qerko:lightning`) and we just experiment for now.

If we end-up with many companion apps like this we can re-think.

I see these companion apps as a bootstrap really. We'll see an increasing amount of native support across PoS providers (although this doesn;t really apply to Qerko given their approach to at table ordering and having the customer manage taking payment).

**Comment by Berserk-21** [comment]
Agree with @bubelov about scaling. Tags also are not super user friendly, better if we don't overcomplicate them with specific payment providers tags

**Comment by bubelov** [comment]
> I'm not a fan of payment:lightning:requires_companion_app = yes or payment:lightning:direct = false as they don't give me enough information as a customer.

Those are base tags, you can always add generic disclaimer tags which we can look into if we encounter a base tag, here is an example:

`payment:lightning:requires_companion_app` = yes
`payment:lightning:companion_app_url` = https://www.qerko.com

The African groceries can be tagged in the same way, but with a different URL. Hard-coding companion apps and inventing unique tagging schemes for them just complicates things and blocks new integrations since they would have to wait till we update all of our front end apps, and we'll need to repeat this process every time.

The tags listed above are sufficient to show the warnings and redirect the users to the webpages with further details.

**Comment by dadofsambonzuki** [comment]
I like the `companion_app_url` idea and this is much more scalable.

**Comment by hynek-jina** [comment]
I think it is solved, right? Or what is the next step?

We are almost ready to upload the Qerko businesses.
I assume we would add these tags:
- `currency:XBT`=yes
- `check_date`=YYYY-MM-DD
- `payment:lightning:requires_companion_app` = yes
- `payment:lightning:companion_app_url` = https://www.qerko.com/

**Comment by dadofsambonzuki** [comment]
LFG

**Comment by bubelov** [comment]
It looks like there are no timestamp tags so all of those places are displayed as unverified, is it intentional?
![Screenshot from 2023-08-02 10-43-34](https://github.com/teambtcmap/btcmap-data/assets/1189245/547774af-3c05-4c41-a683-5ad670a36419)

**Comment by bubelov** [comment]
I think you can safely add `check_date` = `yyyy-mm-dd` for the places which you created, and `check_date:currency:XBT` = `yyyy-mm-dd` for the existing places which you modified. The former tag assumes that you checked all the tags and the the latter one is scoped to Bitcoin related tags only.

**Comment by anton-isidore** [comment]
@hynek-jina @pepakriz awesome thanks everyone for great work 🥳

**Comment by bubelov** [comment]
Ignoring the OSM mods is not helping: https://www.openstreetmap.org/changeset/139308759

You're risking being banned and you changes may be reverted.

**Comment by bubelov** [comment]
The changes have been reverted: https://www.openstreetmap.org/changeset/139507782#map=8/49.793/15.478

Feel free to re-open this issue if you have any additional questions.

**Comment by dadofsambonzuki** [comment]
I think we should keep this open if the plan is to properly engage with the local mods.

If not, then sure we can close.

**Comment by bubelov** [comment]
The problem is: we aren't local mods and this issue can give people a wrong impression that it's OK to import the places since "the BTC Map guys are aware of it and they aren't against it", which will move the spotlight on us and strain our relationship with OSM. We have no authority to decide on anything mentioned above, we can only point to the rules and offer our feedback, which we did. I think this conversation should continue on https://community.openstreetmap.org/

**Comment by bubelov** [comment]
FYI: Some locals are disputing the trustworthiness of Querko dataset:

https://www.openstreetmap.org/changeset/139395653#map=19/50.10110/14.44432

**Comment by hynek-jina** [comment]
Thank you.. I'm off and will get back to it in two weeks..

**Comment by dadofsambonzuki** [comment]
I see there is more Qerko activity:

https://btcmap.org/tagger/19820033

Have local OSM mods been engaged prior to these changes as per the guidance?

If not, expect these to be reverted. Unplanned changes associated with BTC Map do not help with general community relations!

**Comment by anton-isidore** [comment]
@hynek-jina  /@pepakriz are there any news? Have you got chance to figure out the situation with local OSM guys?

**Comment by hynek-jina** [comment]
Hi we were not able to get clear confirmation.
Meanwhile we have [documented](https://wiki.openstreetmap.org/wiki/Import/Qerko_Businesses) these updates and will proceed them at small batches first.

**Comment by bubelov** [comment]
Not sure if it's intentional, but it looks like your sync script is not closing the change sets it makes. It's better not to wait for an automatic forced close (24h) since it delays any discussions. 

Just wanted to remind you to include the verification date tags, as it helps the users to assess the reliability of the data they see on the map.

![Screenshot from 2023-10-23 13-54-25](https://github.com/teambtcmap/btcmap-data/assets/1189245/2ce7774e-19d3-427d-b3c8-92c1bb3fa94f)

**Comment by pepakriz** [comment]
@bubelov thank you for the comment. I'll add a closing change set call. Currently, it's missing because I didn't understand the reason why I should include it. All changes were displayed on the map immediately

**Comment by pepakriz** [comment]
@bubelov I've created two more changesets with a 'close' call (https://www.openstreetmap.org/changeset/143012886, https://www.openstreetmap.org/changeset/143013454). Additionally, I've included 'check_date:currency:XBT' based on the most recent BTC payment.

**Comment by bubelov** [comment]
I guess we can, I see a regular Qerko bot activity in small batches. Feel free to re-open if there is anything else to discuss
